File: lib/models/plex_filter.dart
import 'package:json_annotation/json_annotation.dart';

part 'plex_filter.g.dart';

@JsonSerializable()
class PlexFilter {
  @JsonKey(defaultValue: '')
  final String filter;
  @JsonKey(defaultValue: 'string')
  final String filterType;
  @JsonKey(defaultValue: '')
  final String key;
  @JsonKey(defaultValue: '')
  final String title;
  @JsonKey(defaultValue: 'filter')
  final String type;

  PlexFilter({
    required this.filter,
    required this.filterType,
    required this.key,
    required this.title,
    required this.type,
  });

  factory PlexFilter.fromJson(Map<String, dynamic> json) => _$PlexFilterFromJson(json);

  Map<String, dynamic> toJson() => _$PlexFilterToJson(this);
}

@JsonSerializable(includeIfNull: false)
class PlexFilterValue {
  @JsonKey(defaultValue: '')
  final String key;
  @JsonKey(defaultValue: '')
  final String title;
  final String? type;

  PlexFilterValue({required this.key, required this.title, this.type});

  factory PlexFilterValue.fromJson(Map<String, dynamic> json) => _$PlexFilterValueFromJson(json);

  Map<String, dynamic> toJson() => _$PlexFilterValueToJson(this);
}

File: apple-tv/VibeStream/Models/PlexFilter.swift
import Foundation

struct PlexFilter: Identifiable, Hashable {
    let filter: String
    let filterType: String
    let key: String
    let title: String
    let type: String

    var id: String { key }

    static func from(json: [String: Any]) -> PlexFilter? {
        guard let filter = json["filter"] as? String,
              let filterType = json["filterType"] as? String ?? json["type"] as? String,
              let key = json["key"] as? String,
              let title = json["title"] as? String else {
            return nil
        }
        return PlexFilter(
            filter: filter,
            filterType: filterType,
            key: key,
            title: title,
            type: json["type"] as? String ?? filterType
        )
    }
}

struct PlexFilterValue: Identifiable, Hashable {
    let key: String
    let title: String
    var type: String?

    var id: String { key }

    static func from(json: [String: Any]) -> PlexFilterValue? {
        guard let key = json["key"] as? String ?? json["fastKey"] as? String,
              let title = json["title"] as? String else {
            return nil
        }
        return PlexFilterValue(key: key, title: title, type: json["type"] as? String)
    }
}
